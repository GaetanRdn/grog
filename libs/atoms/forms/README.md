# atoms-forms

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test atoms-forms` to execute the unit tests.
