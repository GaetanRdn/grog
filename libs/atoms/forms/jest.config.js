module.exports = {
  displayName: 'atoms-forms',
  preset: '../../../jest.preset.js',
  setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
      stringifyContentPathRegex: '\\.(html|svg)$',
    },
  },
  collectCoverage: true,
  coverageDirectory: '../../../coverage/libs/atoms/forms',
  transform: {
    '^.+\\.(ts|js|html)$': 'jest-preset-angular',
  },
  snapshotSerializers: [
    'jest-preset-angular/build/serializers/no-ng-attributes',
    'jest-preset-angular/build/serializers/ng-snapshot',
    'jest-preset-angular/build/serializers/html-comment',
  ],
};
