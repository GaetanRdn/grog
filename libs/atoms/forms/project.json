{
  "projectType": "library",
  "root": "libs/atoms/forms",
  "sourceRoot": "libs/atoms/forms/src",
  "prefix": "gro",
  "targets": {
    "build": {
      "executor": "@nrwl/angular:ng-packagr-lite",
      "outputs": ["dist/libs/atoms/forms"],
      "options": {
        "project": "libs/atoms/forms/ng-package.json"
      },
      "configurations": {
        "production": {
          "tsConfig": "libs/atoms/forms/tsconfig.lib.prod.json"
        },
        "development": {
          "tsConfig": "libs/atoms/forms/tsconfig.lib.json"
        }
      },
      "defaultConfiguration": "production"
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/libs/atoms/forms"],
      "options": {
        "jestConfig": "libs/atoms/forms/jest.config.js",
        "passWithNoTests": true
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "options": {
        "lintFilePatterns": [
          "libs/atoms/forms/src/**/*.ts",
          "libs/atoms/forms/src/**/*.html"
        ]
      }
    },
    "storybook": {
      "executor": "@nrwl/storybook:storybook",
      "options": {
        "uiFramework": "@storybook/angular",
        "port": 4400,
        "config": {
          "configFolder": "libs/atoms/forms/.storybook"
        },
        "projectBuildConfig": "atoms-forms"
      },
      "configurations": {
        "ci": {
          "quiet": true
        }
      }
    },
    "build-storybook": {
      "executor": "@nrwl/storybook:build",
      "outputs": ["{options.outputPath}"],
      "options": {
        "uiFramework": "@storybook/angular",
        "outputPath": "dist/storybook/atoms-forms",
        "config": {
          "configFolder": "libs/atoms/forms/.storybook"
        },
        "projectBuildConfig": "atoms-forms"
      },
      "configurations": {
        "ci": {
          "quiet": true
        }
      }
    }
  },
  "tags": [],
  "implicitDependencies": ["styles", "tests", "types"]
}
