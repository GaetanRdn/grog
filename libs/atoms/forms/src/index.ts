export * from './lib/input/input.directive';
export * from './lib/autocomplete/autocomplete.component';
export * from './lib/label/label.component';
export * from './lib/checkbox/checkbox.component';
export * from './lib/select/select.component';
