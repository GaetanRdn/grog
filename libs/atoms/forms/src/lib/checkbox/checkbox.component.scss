@use '../../../../../styles/src/lib/utils';

@mixin gro-checkbox($config) {
  gro-checkbox {
    @include utils.set-typography($config, input);
    display: inline-flex;
    align-items: center;
    cursor: pointer;

    &:not(.gro-disabled, .gro-readonly) {
      @include utils.reverse-color-values($config, primary);

      &:hover,
      &:focus {
        @include utils.reverse-color-values($config, primary-dark);
      }
    }

    &.gro-disabled,
    &.gro-readonly {
      @include utils.reverse-color-values($config, disabled);
    }

    input[type='checkbox' i] {
      height: 1rem;
      width: 1rem;
      cursor: inherit;
    }
  }
}
