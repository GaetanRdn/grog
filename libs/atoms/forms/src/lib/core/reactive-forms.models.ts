export type OnChangeFn<ValueType> = (value: ValueType | null) => void;
export type OnTouchedFn = () => void;
