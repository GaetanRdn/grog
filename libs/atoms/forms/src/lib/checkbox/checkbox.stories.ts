import { FormControl, ReactiveFormsModule } from '@angular/forms';
import { action } from '@storybook/addon-actions';
import { Meta, Story } from '@storybook/angular';
import { CheckboxComponent } from '@grorg/atoms-forms';

export default {
  title: 'atoms/forms/checkbox',
  component: CheckboxComponent,
  args: {
    ngContent: 'Check me',
  },
  argTypes: {
    value: { control: { type: 'text' } },
  },
  parameters: {
    jest: ['checkbox.component.spec.ts'],
  },
} as Meta<CheckboxComponent<string>>;

const basicTemplate: Story<CheckboxComponent<string>> = (
  args: CheckboxComponent<string>
) => ({
  template: `<adr-checkbox [value]="value" [checked]="checked" [disabled]="disabled" [readOnly]="readOnly" (valueChange)="valueChange($event)">{{ ngContent }}</adr-checkbox>`,
  props: {
    ...args,
    valueChange: action('valueChange'),
  },
});

export const basic = basicTemplate.bind({});
basic.args = {
  checked: false,
  disabled: false,
  readOnly: false,
  value: 'test',
};

const reactiveTemplate: Story<CheckboxComponent<string>> = (
  args: CheckboxComponent<string>
) => ({
  moduleMetadata: {
    imports: [ReactiveFormsModule],
  },
  template: `<adr-checkbox [value]="value" [readOnly]="readOnly" [formControl]="control">{{ ngContent }}</adr-checkbox>`,
  props: {
    ...args,
    control: new FormControl(args.value),
  },
});

export const reactiveForms = reactiveTemplate.bind({});
reactiveForms.args = {
  readOnly: false,
  value: 'test',
};
