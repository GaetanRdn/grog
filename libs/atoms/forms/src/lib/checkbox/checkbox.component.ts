import { CommonModule } from '@angular/common';
import {
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  EventEmitter,
  HostBinding,
  HostListener,
  Input,
  NgModule,
  Output,
} from '@angular/core';
import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
import { AutoUnsubscribe, CoerceBoolean } from '@grorg/decorators';
import { OnChangeFn, OnTouchedFn } from '../core/reactive-forms';

@Component({
  selector: 'gro-checkbox',
  // s
  host: {
    '[attr.checked]': 'checked || null',
    '[attr.readonly]': 'readOnly || null',
    '[attr.disabled]': 'disabled || null',
    '[class.adr-checked]': 'checked || null',
    '[class.adr-readonly]': 'readOnly || null',
    '[class.adr-disabled]': 'disabled || null',
  },
  template: `<input
      type="checkbox"
      [value]="value"
      [readOnly]="readOnly || null"
      [disabled]="disabled || null"
      [checked]="checked || null"
    />
    <span class="adr-checkbox-label"><ng-content></ng-content></span>`,
  providers: [
    { provide: NG_VALUE_ACCESSOR, useExisting: CheckboxComponent, multi: true },
  ],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
@AutoUnsubscribe()
export class CheckboxComponent<ValueType> implements ControlValueAccessor {
  // TIPS: HTMLInputElement.indeterminate = true;

  @Input()
  @HostBinding('attr.value')
  public value: ValueType | null = null;

  @Input()
  @CoerceBoolean()
  public checked?: boolean;

  @Input()
  @CoerceBoolean()
  public readOnly?: boolean;

  @Input()
  @CoerceBoolean()
  public disabled?: boolean;

  @Output()
  public readonly valueChange: EventEmitter<ValueType | null> = new EventEmitter<ValueType | null>();

  constructor(private _changeDetectorRef: ChangeDetectorRef) {}

  public writeValue(obj: ValueType): void {
    this.checked = obj !== null && obj !== undefined;
  }

  public registerOnChange(fn: OnChangeFn<ValueType>): void {
    this._onChange = fn;
  }

  public registerOnTouched(fn: OnTouchedFn): void {
    this._onTouched = fn;
  }

  public setDisabledState(isDisabled: boolean): void {
    this.disabled = isDisabled;
    this._changeDetectorRef.markForCheck();
  }

  /**
   * @internal private usage
   */
  @HostListener('click')
  public onChange(): void {
    if (!this.readOnly && !this.disabled) {
      this._onTouched();
      this.checked = !this.checked;
      const emittedValue: ValueType | null = this.checked ? this.value : null;
      this.valueChange.emit(emittedValue);
      this._onChange(emittedValue);
    }
  }

  protected _onChange: OnChangeFn<ValueType> = (): void => {
    // default if no ngControl
  };

  protected _onTouched: () => void = (): void => {
    // default if no ngControl
  };
}

@NgModule({
  declarations: [CheckboxComponent],
  exports: [CheckboxComponent],
  imports: [CommonModule],
})
export class CheckboxModule {}
