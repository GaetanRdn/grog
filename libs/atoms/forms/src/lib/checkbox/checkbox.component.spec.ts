import { Component, OnDestroy } from '@angular/core';
import { TestBed } from '@angular/core/testing';
import { FormControl, ReactiveFormsModule } from '@angular/forms';
import { CheckboxComponent } from './checkbox.component';
import { TemplateLookup } from '@grorg/tests';

describe('CheckboxComponent', () => {
  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [
        CheckboxComponent,
        BasicHostComponent,
        ReadonlyHostComponent,
        DisabledHostComponent,
        ReactiveFormHostComponent,
      ],
      imports: [ReactiveFormsModule],
      providers: [CheckboxComponent],
    });
  });

  describe('Basic checkbox', () => {
    let templateLookup: TemplateLookup<BasicHostComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup<BasicHostComponent>(
        TestBed.createComponent(BasicHostComponent)
      );
    });

    test.each(['val1', '2'])('should create with value %s', (value: string) => {
      // GIVEN
      templateLookup.hostComponent.checkboxValue = value;

      // WHEN
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.query<HTMLInputElement>('input').value).toEqual(
        value
      );
    });

    test('should emit value on check', () => {
      // GIVEN
      templateLookup.hostComponent.checkboxValue = 'val';
      templateLookup.hostComponent.checked = false;
      expect(templateLookup.hostComponent.currentValue).toBeNull();
      templateLookup.detectChanges();

      // WHEN
      templateLookup.query<HTMLInputElement>(CheckboxComponent).click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.hostComponent.currentValue).toEqual('val');
    });

    test('should emit null on uncheck', () => {
      // GIVEN
      templateLookup.hostComponent.checkboxValue = 'val';
      templateLookup.hostComponent.currentValue = 'val';
      templateLookup.hostComponent.checked = true;
      templateLookup.detectChanges();

      // WHEN
      templateLookup.query<HTMLInputElement>(CheckboxComponent).click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.hostComponent.currentValue).toBeNull();
    });

    test('check OnDestroy', () => {
      // GIVEN
      const checkbox: CheckboxComponent<string> = templateLookup
        .get('input')
        .injector.get(CheckboxComponent);
      jest.spyOn(checkbox.valueChange, 'unsubscribe');

      // WHEN
      (checkbox as unknown as OnDestroy).ngOnDestroy();

      // THEN
      expect(checkbox.valueChange.unsubscribe).toBeCalled();
    });
  });

  describe('Readonly checkbox', () => {
    let templateLookup: TemplateLookup<ReadonlyHostComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup<ReadonlyHostComponent>(
        TestBed.createComponent(ReadonlyHostComponent)
      );

      templateLookup.detectChanges();
    });

    test('check does not emit value', () => {
      // GIVEN
      const checkbox: CheckboxComponent<string> = templateLookup
        .get('input')
        .injector.get(CheckboxComponent);
      jest.spyOn(checkbox.valueChange, 'emit');

      // WHEN
      templateLookup.query<HTMLInputElement>(CheckboxComponent).click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(checkbox.valueChange.emit).not.toBeCalled();
    });
  });

  describe('Disabled checkbox', () => {
    let templateLookup: TemplateLookup<DisabledHostComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup<DisabledHostComponent>(
        TestBed.createComponent(DisabledHostComponent)
      );

      templateLookup.detectChanges();
    });

    test('check does not emit value', () => {
      // GIVEN
      const checkbox: CheckboxComponent<string> = templateLookup
        .get('input')
        .injector.get(CheckboxComponent);
      jest.spyOn(checkbox.valueChange, 'emit');

      // WHEN
      templateLookup.query<HTMLInputElement>(CheckboxComponent).click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(checkbox.valueChange.emit).not.toBeCalled();
    });
  });

  describe('ReactiveForm checkbox', () => {
    let templateLookup: TemplateLookup<ReactiveFormHostComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup<ReactiveFormHostComponent>(
        TestBed.createComponent(ReactiveFormHostComponent)
      );

      templateLookup.detectChanges();
    });

    test('should create checked', () => {
      // GIVEN
      const checkbox: CheckboxComponent<string> = templateLookup
        .get('input')
        .injector.get(CheckboxComponent);

      // WHEN
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(checkbox.value).toEqual(
        templateLookup.hostComponent.control.value
      );
    });

    test('should create unchecked', () => {
      // GIVEN
      const checkbox: CheckboxComponent<string> = templateLookup
        .get('input')
        .injector.get(CheckboxComponent);
      templateLookup.hostComponent.control.setValue(null);

      // WHEN
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(checkbox.value).toEqual('val');
    });

    test('check is touched on change', () => {
      // WHEN
      templateLookup.query<HTMLInputElement>(CheckboxComponent).click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('check control value on change', () => {
      // WHEN
      templateLookup.query<HTMLInputElement>(CheckboxComponent).click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.hostComponent.control.value).toBeNull();
    });

    test('should create disabled', () => {
      // GIVEN
      templateLookup.hostComponent.control.disable();

      // WHEN
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });
});

@Component({
  template: `<adr-checkbox
    value="{{ checkboxValue }}"
    (valueChange)="currentValue = $event"
    [checked]="checked"
    >Basic</adr-checkbox
  >`,
})
class BasicHostComponent {
  public checkboxValue!: any;

  public currentValue: any = null;

  public checked = false;
}

@Component({
  template: ` <adr-checkbox [value]="'val'" readOnly>ReadOnly</adr-checkbox>`,
})
class ReadonlyHostComponent {}

@Component({
  template: `<adr-checkbox [value]="'val'" disabled>Disabled</adr-checkbox>`,
})
class DisabledHostComponent {}

@Component({
  template: `<adr-checkbox [value]="'val'" [formControl]="control"
    >Reactive</adr-checkbox
  >`,
})
class ReactiveFormHostComponent {
  public control: FormControl = new FormControl('val');
}
