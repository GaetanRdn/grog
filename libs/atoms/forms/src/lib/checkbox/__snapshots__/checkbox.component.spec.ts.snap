// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxComponent Basic checkbox should create with value 2 1`] = `
<gro-checkbox
  value="2"
>
  <input
    type="checkbox"
    value="2"
  />
  <span
    class="adr-checkbox-label"
  >
    Basic 
  </span>
</gro-checkbox>
`;

exports[`CheckboxComponent Basic checkbox should create with value val1 1`] = `
<gro-checkbox
  value="val1"
>
  <input
    type="checkbox"
    value="val1"
  />
  <span
    class="adr-checkbox-label"
  >
    Basic 
  </span>
</gro-checkbox>
`;

exports[`CheckboxComponent Basic checkbox should emit null on uncheck 1`] = `
<gro-checkbox
  value="val"
>
  <input
    type="checkbox"
    value="val"
  />
  <span
    class="adr-checkbox-label"
  >
    Basic 
  </span>
</gro-checkbox>
`;

exports[`CheckboxComponent Basic checkbox should emit value on check 1`] = `
<gro-checkbox
  checked="true"
  class="adr-checked"
  value="val"
>
  <input
    type="checkbox"
    value="val"
  />
  <span
    class="adr-checkbox-label"
  >
    Basic 
  </span>
</gro-checkbox>
`;

exports[`CheckboxComponent Disabled checkbox check does not emit value 1`] = `
<gro-checkbox
  class="adr-disabled"
  disabled="true"
  value="val"
>
  <input
    disabled=""
    type="checkbox"
    value="val"
  />
  <span
    class="adr-checkbox-label"
  >
    Disabled
  </span>
</gro-checkbox>
`;

exports[`CheckboxComponent ReactiveForm checkbox check is touched on change 1`] = `
<gro-checkbox
  class="ng-valid ng-touched ng-dirty"
  value="val"
>
  <input
    type="checkbox"
    value="val"
  />
  <span
    class="adr-checkbox-label"
  >
    Reactive 
  </span>
</gro-checkbox>
`;

exports[`CheckboxComponent ReactiveForm checkbox should create checked 1`] = `
<gro-checkbox
  checked="true"
  class="adr-checked ng-untouched ng-pristine ng-valid"
  value="val"
>
  <input
    type="checkbox"
    value="val"
  />
  <span
    class="adr-checkbox-label"
  >
    Reactive 
  </span>
</gro-checkbox>
`;

exports[`CheckboxComponent ReactiveForm checkbox should create disabled 1`] = `
<gro-checkbox
  checked="true"
  class="adr-checked ng-untouched ng-pristine adr-disabled"
  disabled="true"
  value="val"
>
  <input
    disabled=""
    type="checkbox"
    value="val"
  />
  <span
    class="adr-checkbox-label"
  >
    Reactive 
  </span>
</gro-checkbox>
`;

exports[`CheckboxComponent ReactiveForm checkbox should create unchecked 1`] = `
<gro-checkbox
  class="ng-untouched ng-pristine ng-valid"
  value="val"
>
  <input
    type="checkbox"
    value="val"
  />
  <span
    class="adr-checkbox-label"
  >
    Reactive 
  </span>
</gro-checkbox>
`;

exports[`CheckboxComponent Readonly checkbox check does not emit value 1`] = `
<gro-checkbox
  class="adr-readonly"
  readonly="true"
  value="val"
>
  <input
    readonly=""
    type="checkbox"
    value="val"
  />
  <span
    class="adr-checkbox-label"
  >
    ReadOnly
  </span>
</gro-checkbox>
`;
